src/vm/internal/otable.cc

```diff
@@ -1,271 +1,148 @@
-
-  DEBUG_CHECK1(s != ob, "Remove object \"/%s\": found a different object!", ob->obname);
-
-  obj_table[h] = ob->next_hash;
-  ob->next_hash = 0;
-  objs_in_table--;
-
-  base = _basename(ob->obname, &dummy);
-  ch = ObjHash(base);
-  FREE_MSTR(base);
-  s = ch_table[ch];
-  while (s && s != ob) {
-    t = s;
-    s = s->next_ch_hash;
-  }
-  if (!s) {
-    fatal("object not found in children list");
-  }
-  if (!t) {
-    ch_table[ch] = s->next_ch_hash;
-  } else {
-    t->next_ch_hash = s->next_ch_hash;
-  }
-  ob->next_ch_hash = 0;
-  return;
+//Attempt to find the object with name key in the object table and remove it
+//from both the objects map and the children map if found. Returns true if found,
+//otherwise false.
+//TODO: generalize this code by removing dependency on field obname.
+bool ObjectTable::remove(Key const & key) {
+    
+    auto it1 = objects_.find(key);
+    if( it1 != objects_.end() ) {
+        objects_.erase(it1);
+        //guaranteed to exist if object exists
+        //std::assert( children_.find( basename(key)) != children_.end() );
+        auto it2 = children_.find( basename(key) );
+        auto it3 = find_if( it2->second.begin(), it2->second.end(), [&key](Value v) -> bool { return v->obname == key; } );
+        if(it3 != it2->second.end() ) {
+            it2->second.erase(it3);
+        }
+        else {
+            return false;
+        }
+        if( it2->second.size() == 0 ) {
+            children_.erase(it2);
+        }
+        return true;
+    }
+    else {
+        return false;
+    }
 }
 
-/*
- * Lookup an object in the hash table; if it isn't there, return null.
- * This is only different to find_object_n in that it collects different
- * stats; more finds are actually done than the user ever asks for.
- */
-
-static long user_obj_lookups = 0, user_obj_found = 0;
-
-object_t *lookup_object_hash(const char *s) {
-  object_t *ob = find_obj_n(s);
+//write some information about memory usage and other statistics of the object table to out_buffer_t.
+//TODO: remove dependency on outbuffer_t here
+#ifndef TESTING
+int ObjectTable::showStatus(outbuffer_t *out, int verbose) {
+
+    std::stringstream ss;
+    if (verbose == 1) {
+        ss <<  "Object name hash table status:" << std::endl;
+        ss <<  "Object name hash table status:" << std::endl;
+        ss << "------------------------------" << std::endl;
+        ss << "Elements:        " << objects_.size() << std::endl;
+        ss << "Memory(bytes):     " << objects_.size() * sizeof(Value) << std::endl;
+        ss << "Bucket count:    " << objects_.bucket_count() << std::endl;
+        ss << "Load factor:     " << objects_.load_factor() << std::endl;
+
+        outbuf_add(out, ss.str().c_str());
+      }
 
-  user_obj_lookups++;
-  if (ob) {
-    user_obj_found++;
-  }
-  return (ob);
+    if (!verbose) {
+        ss << "Memory used(bytes):     " << objects_.size() * sizeof(Value) << std::endl;
+        outbuf_add(out, ss.str().c_str());
```

Comment:
this if() case can be changed to a "else{]"

Also, ```outbuf_add(..)``` could be moved out so you don't have to repeate it .

Reply:
Fixed.

Reply:
Oversight on my part was just copying the old code with looking for simplifications. Thanks.